src/backends/torch/torchinputconns.cc

```diff
@@ -750,28 +753,146 @@ namespace dd
-    // , _test_dataset
-    std::vector<std::vector<CSVline>> *data;
-    if (use_csvtsdata_test)
-      data = &this->_csvtsdata_test;
+    std::vector<at::Tensor> data_sequence;
+    if (_fnames.size() > static_cast<unsigned int>(vecindex))
+      _ids.push_back(_fnames[vecindex] + " #" + std::to_string(tstart) + "_"
+                     + std::to_string(tstart + _timesteps - 1));
+    for (long int ti = tstart; ti < tstart + _timesteps - _forecast; ++ti)
+      {
+        std::vector<double> datavec;
+        for (int di = 0; di < this->_datadim; ++di)
+          datavec.push_back(seq[ti]._v[di]);
+        at::Tensor data
+            = torch::from_blob(&datavec[0], at::IntList{ _datadim },
+                               torch::kFloat64)
+                  .clone()
+                  .to(torch::kFloat32);
+        data_sequence.push_back(data);
+      }
+    at::Tensor dst = torch::stack(data_sequence);
+
+    if (static_cast<int>(seq.size()) >= _timesteps + tstart)
+      {
+        std::vector<at::Tensor> pred_sequence;
+        for (long int ti = tstart + _timesteps - _forecast;
+             ti < tstart + _timesteps; ++ti)
+          {
+            std::vector<double> predvec;
+            for (int di = 0; di < this->_datadim; ++di)
+              predvec.push_back(seq[ti]._v[di]);
+            at::Tensor pred
+                = torch::from_blob(&predvec[0], at::IntList{ _datadim },
+                                   torch::kFloat64)
+                      .clone()
+                      .to(torch::kFloat32);
+            pred_sequence.push_back(pred);
+          }
+        at::Tensor pst = torch::stack(pred_sequence);
+        dataset.add_batch({ dst }, { pst });
+      }
+    else // we are in inference mode, not forecast available
+      dataset.add_batch({ dst }, {});
+  }
+
+  void CSVTSTorchInputFileConn::discard_warn(
+      int vecindex, unsigned int seq_size,
+      const std::vector<std::vector<CSVline>> *data)
+  {
+    std::string errmsg = "data does not contains enough timesteps, "
+                         "discarding (seq_size:"
+                         + std::to_string(seq_size)
+                         + " _timesteps:" + std::to_string(_timesteps);
+    if (data == &this->_csvtsdata_test)
+      {
+        if (static_cast<unsigned int>(vecindex) < _test_fnames.size())
+          errmsg = "file " + _test_fnames[vecindex]
+                   + " does not contains enough timesteps, "
+                     "discarding";
+      }
     else
-      data = &this->_csvtsdata;
+      {
+        if (static_cast<unsigned int>(vecindex) < _fnames.size())
+          errmsg = "file " + _fnames[vecindex]
+                   + " does not contains enough timesteps, "
+                     "discarding";
+      }
+    _tilogger->warn(errmsg);
+  }
+
+  void CSVTSTorchInputFileConn::fill_dataset_forecast(
+      TorchDataset &dataset, const std::vector<std::vector<CSVline>> *data)
```

Comment:
It's better to pass references as parameter

Reply:
ok, done